components/script/dom/servoparser/html.rs

```diff
@@ -115,79 +115,132 @@ unsafe impl JSTraceable for HtmlTokenizer<TreeBuilder<JS<Node>, Sink>> {
+struct SerializationIterator {
+    stack: Vec<TreeIterator>,
+}
 
-            (ChildrenOnly, NodeTypeId::Document(_)) => {
-                for handle in node.children() {
-                    (&*handle).serialize(serializer, IncludeNode)?;
-                }
-                Ok(())
-            },
+impl SerializationIterator {
+    fn new(node: &Node) -> SerializationIterator {
+        SerializationIterator {
+            stack: vec![node.traverse_preorder()],
+        }
+    }
+}
 
-            (ChildrenOnly, _) => Ok(()),
+impl Iterator for SerializationIterator {
+    type Item = Root<Node>;
 
-            (IncludeNode, NodeTypeId::DocumentType) => {
-                let doctype = node.downcast::<DocumentType>().unwrap();
-                serializer.write_doctype(&doctype.name())
-            },
+    fn next(&mut self) -> Option<Root<Node>> {
+        if self.stack.len() == 0 { return None; }
 
-            (IncludeNode, NodeTypeId::CharacterData(CharacterDataTypeId::Text)) => {
-                let cdata = node.downcast::<CharacterData>().unwrap();
-                serializer.write_text(&cdata.data())
-            },
+        if let Some(n) = self.stack.last_mut().unwrap().next() {
+            // https://github.com/w3c/DOM-Parsing/issues/1
+            if let Some(tpl) = n.downcast::<HTMLTemplateElement>() {
+                // We need to visit the contents of the template from left to right.
+                // Since we always work from the top of the stack down, push the iterators in reverse order.
+                for c in tpl.Content().upcast::<Node>().rev_children() {
+                    self.stack.push(c.traverse_preorder());
+                }
+            }
+            return Some(n);
+        } else {
+            self.stack.pop();
+        }
 
-            (IncludeNode, NodeTypeId::CharacterData(CharacterDataTypeId::Comment)) => {
-                let cdata = node.downcast::<CharacterData>().unwrap();
-                serializer.write_comment(&cdata.data())
-            },
+        None
+    }
+}
 
-            (IncludeNode, NodeTypeId::CharacterData(CharacterDataTypeId::ProcessingInstruction)) => {
-                let pi = node.downcast::<ProcessingInstruction>().unwrap();
-                let data = pi.upcast::<CharacterData>().data();
-                serializer.write_processing_instruction(&pi.target(), &data)
-            },
+impl<'a> Serialize for &'a Node {
+    fn serialize<S: Serializer>(&self, serializer: &mut S,
+                                traversal_scope: TraversalScope) -> io::Result<()> {
+        let node = *self;
 
-            (IncludeNode, NodeTypeId::DocumentFragment) => Ok(()),
+        // As we encounter elements, we push them to this stack with a count of their children.
+        // Every time through the following loop, we decrement the child count of the topmost element.
+        // When it gets to 0, it's time to close the node.
+        let mut open_stack: Vec<(Root<Node>, u32)> = vec![];
```

Comment:
It would be nice if we could roll this stack into the iterator, rather than have it explicit.